src/session-pool.js

```diff
@@ -75,874 +129,662 @@ const READWRITE = 'readwrite';
-  const formatted = trace.slice(2).map(function(t) {
-    return `    at ${t.getFunctionName() ||
-      t.getMethodName()} (${t.getFileName()}:${t.getLineNumber()}:${t.getColumnNumber()})`;
-  });
-
-  return `Session leak detected!\n${formatted.join('\n')}`;
-};
-
-/**
- * Returns the number of available sessions.
- *
- * @return {number}
- */
-SessionPool.prototype.available = function() {
-  return this.reads_.length + this.writes_.length;
-};
-
-/**
- * Returns the number of borrowed sessions.
- *
- * @return {number}
- */
-SessionPool.prototype.borrowed = function() {
-  return this.borrowed_.length;
-};
+  get borrowed() {
+    return this._inventory.borrowed.size;
+  }
 
-/**
- * Closes and the pool.
- *
- * @emits SessionPool#close
- * @returns {Promise}
- */
-SessionPool.prototype.close = function() {
-  const self = this;
-  const sessions = this.reads_.concat(this.writes_, this.borrowed_);
+  /**
+   * Flag to determine if Pool is full.
+   * @type {boolean}
+   */
+  get isFull() {
+    return this.size >= this.options.max;
+  }
 
-  this.isOpen = false;
+  /**
+   * Total number of read sessions.
+   * @type {number}
+   */
+  get reads() {
+    const {readonly, borrowed} = this._inventory;
+    const available = readonly.length;
+    const used = Array.from(borrowed).filter(
+      session => session.type === READONLY
+    ).length;
+
+    return available + used;
+  }
 
-  this.reads_ = [];
-  this.writes_ = [];
-  this.borrowed_ = [];
+  /**
+   * Total size of pool.
+   * @type {number}
+   */
+  get size() {
+    return this.available + this.borrowed;
+  }
 
-  this.stopHouseKeeping_();
+  /**
+   * Total number of write sessions.
+   * @type {number}
+   */
+  get writes() {
+    const {readwrite, borrowed} = this._inventory;
+    const available = readwrite.length;
+    const used = Array.from(borrowed).filter(
+      session => session.type === READWRITE
+    ).length;
+
+    return available + used;
+  }
 
-  this.emit('empty');
-  this.emit('close');
+  /**
+   * @constructor
+   * @param {Database} database The DB instance.
+   * @param {SessionPoolOptions} [options] Configuration options.
+   */
+  constructor(database, options) {
+    super();
 
-  return Promise.all(
-    sessions.map(function(session) {
-      return self.destroySession_(session);
-    })
-  );
-};
+    this.isOpen = false;
+    this.database = database;
+    this.options = Object.assign({}, DEFAULTS, options);
 
-/**
- * Fills the pool with the minimum number of sessions.
- *
- * @return {Promise}
- */
-SessionPool.prototype.fill = function() {
-  const requests = [];
-  let i;
+    if (this.options.writes > 1) {
+      throw new WritePercentError();
+    }
 
-  const minReadWrite = Math.floor(this.options.min * this.options.writes);
-  const neededReadWrite = Math.max(minReadWrite - this.writes_.length, 0);
+    this._inventory = {
+      [READONLY]: [],
+      [READWRITE]: [],
+      borrowed: new Set(),
```

Comment:
curious why a `Set` here but array for the other inventories?

Reply:
We want to maintain a specific order for the available options, making sure to give back the recently used sessions first. Borrowed sessions don't matter, so I went with Set to make life easier.